### ISSUE TITLE
Feature-reref

### ISSUE THREAD

**Issue by georgebuzzell** [NDCLab/baseEEG#30]
Feature-reref
-re-reference the data to the average of all electrodes (“average reference”) using the mne function
-write to annotations file that data were re-referenced to average

**Comment by yanbin-niu** [comment]
```
import matplotlib
import pathlib
import mne
import mne_bids
matplotlib.use('Qt5Agg')

# Read BIDS data
bids_root = pathlib.Path('raw_data/eegmatchingpennies')
bids_path = mne_bids.BIDSPath(subject='05',
                              task='matchingpennies',
                              datatype='eeg',
                              root=bids_root)

raw = mne_bids.read_raw_bids(bids_path)

# set reference
raw.load_data()
raw_avg_ref = raw.copy().set_eeg_reference(ref_channels='average')
```

1. for average reference, do we need to add back the on-line reference first?
2. it seems that "Creating the average reference as a projector" is recommended (https://mne.tools/stable/auto_tutorials/preprocessing/plot_55_setting_eeg_reference.html#setting-average-reference), but I have no idea about anything about projector. if it is actually preferable,  we can change to this way.
@georgebuzzell

**Comment by georgebuzzell** [comment]
@yanbin-niu In response to adding back in the on-line ref first, great question. Yes, definitely. I did not check this example dataset yet, do you know what the online ref was for this data?

For #2, I am not familiar with computing avg ref as a projector. I will look into this. However, for now, the method you use seems more than sufficient, as that is bar far the standard approach.

**Comment by georgebuzzell** [comment]
Thanks!!

**Comment by yanbin-niu** [comment]
@georgebuzzell For this example dataset:
    "Manufacturer": "Brain Products",
    "ManufacturersModelName": "BrainAmp DC",
    "CapManufacturer": "Brain Products",
    "CapManufacturersModelName": "actiCAP 64Ch Standard-2",
    "EEGChannelCount": 10,
    "EOGChannelCount": 0,
    "ECGChannelCount": 0,
    "EMGChannelCount": 0,
    "EEGReference": "unipolar, placed on Fz",
    "EEGGround": "placed on Afz",

It seems like Fz is the on-line reference? and it only has 10 electrodes.

**Comment by georgebuzzell** [comment]
@yanbin-niu Thank you! Please let me know if you have questions about how to implement, especially if you have any EEG-content questions (or any questions)!

**Comment by yanbin-niu** [comment]
@georgebuzzell It seems like Fz (on-line reference) was not included in 10 electrodes. In this situation, do we need to add back Fz before we do the re-reference?

**Comment by georgebuzzell** [comment]
@yanbin-niu what do you mean by "not included in 10 electrodes"?

Yes, if the on-line ref was not included, we need to add it back in first. The data for the added in channel (ref) should be all zeros.

**Comment by yanbin-niu** [comment]
@georgebuzzell Some updates:

1. it seems that projector/projection is related to the noisy processing that is commonly used in MEG (thought it also can be used with EEG). I also get an impression that it is relate to PCA. However, I would be completely wrong!! Just get this from https://mne.tools/stable/auto_tutorials/preprocessing/plot_50_artifact_correction_ssp.html#sphx-glr-auto-tutorials-preprocessing-plot-50-artifact-correction-ssp-py

> A projector (abbr. proj), also referred to as Signal Space Projection (SSP), defines a linear operation applied spatially to EEG or MEG data. You can see this as a matrix multiplication that reduces the rank of the data by projecting it to a lower dimensional subspace. Such a projection operator is applied to both the data and the forward operator for source localization. **Note that EEG average referencing can be done using such a projection operator. It is stored in the measurement info in info['projs']**.

Long story short, I guess it is not commonly used with EEG, and we can move forward for now, though the quote does mention it can be used with EEG average referencing. 

2. After adding reference electrode back, the new script goes like:

```
import matplotlib
import pathlib
import mne
import mne_bids

matplotlib.use('Qt5Agg')

# Read BIDS data
bids_root = pathlib.Path('raw_data/eegmatchingpennies')
bids_path = mne_bids.BIDSPath(subject='05',
                              task='matchingpennies',
                              datatype='eeg',
                              root=bids_root)

raw = mne_bids.read_raw_bids(bids_path)

raw.plot()
raw.info
```
<img width="850" alt="image" src="https://user-images.githubusercontent.com/79607547/115134935-32c7a900-a047-11eb-921e-c3e8569aaef4.png">

Here we can see the reference electrode (FZ) is not included. so add it back:

```
# add back reference channel (all zero)
raw.load_data()
raw_new_ref = mne.add_reference_channels(raw, ref_channels=['FZ'])
raw_new_ref.plot()
```

<img width="850" alt="image" src="https://user-images.githubusercontent.com/79607547/115134959-67d3fb80-a047-11eb-9669-21396f8d7d8b.png">

This looks good. FZ is zero. Then do the average referencing:

```
# set reference
raw_avg_ref = raw_new_ref.copy().set_eeg_reference(ref_channels='average')
raw_avg_ref.plot()
```

<img width="850" alt="image" src="https://user-images.githubusercontent.com/79607547/115134997-a964a680-a047-11eb-8f95-ac86b4cad034.png">

**Comment by yanbin-niu** [comment]
@georgebuzzell A question relate to average referencing. I read this early on, https://eeglab.org/tutorials/06_RejectArtifacts/RunICA.html

> When computing average reference on n-channel data, the rank of the data is reduced to n-1. Why? Because the sum of the potential is 0 at all time points, the last channel activity is equal to minus the sum of the others. ICA does not behave well in this (rank-deficient) case.

In both EEGlab and MNE tutorial, I never read something like, when you do the average referencing, you should add online reference back first (I may miss something). Now you mentioned this. Alongside the rank fo the data I quoted above, now I am thinking, since we add back the online reference electrode, and usually the online reference electrode is not included in the raw data,, does this mean we actually do not lose one rank even after doing average referencing? 

I ask this because I think it's related to rank-deficiency. And it seems to be a big deal, since I watched this video, it is mentioned that even when you keep 99% of the variance it’s not the same as using ICA without PCA.(about 7‘10’‘ at https://www.youtube.com/watch?v=2hrYEYSycGI&list=PLXc9qfVbMMN2uDadxZ_OEsHjzcRtlLNxc)
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/79607547/115135791-c7350a00-a04d-11eb-856f-94159f1ff466.png">

If we do not lose one rank even after doing average referencing, does this mean we are not in the "rank-deficient" situation?

**Comment by georgebuzzell** [comment]
@yanbin-niu your code looks excellent; please do a pr!

In terms of rereferincing as a ssp, I do like this idea a lot, and I think we will certainly want to revisit this after we have v.01 of the full pipeline built. For now, I think we should just use the traditional method. But again, I would be very interested in reopening this discussion and likely implementing in the near future, as the ssp would I think be in line with a "lossless" approach to data processing, where we dont change the base data much, just add projection matrices to rereference, remove segments, etc. However, that is all beyond the scope of v.01 of the pipeline. BUT, I need to to a better job of defining the scope of v.01 vs later versions in the roadmap. I will do that soon, since you are already generating so many great ideas and implementations that go beyond v.01 (which is a very good thing!)

For the rank deficiency question you are absolutely correct that adding back in the reference reduces the rank by 1. Essentially, the reference signal contains no unique information (deterministically predicted by all other channels). Thus, it would not be appropriate to add it back in prior to Ica. However, please note in the roadmap that adding in the reference and computing the average reference is done at the end of the preprocessing pipeline, after ica. Thus, there is not an issue here. However, your question was a good one, as that is not clear when we are working on the various features out of order right now. :) I also want to point out that your understanding of the potential issues associated with adding the reference, rank deficiency, pca, and ica are all spot on, and frankly, quite impressive! So, thank you for asking about this! :)

**Comment by F-said** [comment]
Closed with merge of #160
